### PR TITLE
feat!: obfuscation for mysql2, dalli and postgresql as default option for db_statement

### PR DESCRIPTION
**Pull request by xuan-cao-swi**
## Description

Based on [opentelemetry-specification](https://github.com/open-telemetry/opentelemetry-specification/pull/3127) and [discussion](https://github.com/open-telemetry/opentelemetry-ruby-contrib/discussions/442), change the db.statement default sql to sanitized sql

## Test
Tested with basic test suite (e.g. `rake test`)

### CODE REVIEW

**Review by ericmustin — CHANGES_REQUESTED** [review]
Obfuscation adds a significant performance penalty.

We should either omit or include by default. The spec says to omit now if I understand correctly?

**Comment by arielvalentin** [comment]
> Obfuscation adds a significant performance penalty.
> 
> We should either omit or include by default. The spec says to omit now if I understand correctly?

@ericmustin The spec says nothing about the defaults client-side sanitization, AFAIK, other than:

>  [2]: Should be collected by default only if there is sanitization that excludes sensitive information.

I interpret that to mean you should only collect it if the SQL is sanitized. If I extrapolate a bit, then it should never record any unsanitized SQL at all. 

There are some open issues about db statement sanitization in general:

- https://github.com/open-telemetry/opentelemetry-specification/issues/1106
- https://github.com/open-telemetry/opentelemetry-specification/issues/3104
- https://github.com/open-telemetry/opentelemetry-specification/issues/3594

W/R/T `db.operation` it says:

> [3]: When setting this to an SQL keyword, it is not recommended to attempt any client-side parsing of db.statement just to get this property, but it should be set if the operation name is provided by the library being instrumented. If the SQL statement has an ambiguous operation, or performs more than one operation, this value may be omitted.

Given this information I for one prefer to sanitize by default and `omit` if the driver does not have a sanitizer implemented.

**Comment by arielvalentin** [comment]
> The appropriate place to do sanitization is in the collector, or more generally out of process. If your interpretation of the spec is correct(and I think it is), then that would effectively exclude the possibility of doing sql sanitization in the collector because the spec would dictate it can never be exported without being sanitized. This seems flawed to me.
> 
> Additionally we _know_ real examples of the "obfuscate" config option introducing a performance regression in applications at large enterprise organizations which use ruby. Setting it as the default would seem like a poor choice.
> 
> I would prefer omit as the default if we're left with a choice btwn omit and "conditionally obfuscate depending on the particulars of driver options", as I think users will find the lack of standard confusing.

Would it make the most sense to discussion in the spec issues? 

I think the challenge here is there are no OOTB SQL sanitizers in [contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor). If there were, then I would totally be onboard with leaving `include`/`raw` as the default, but since there isn't I still advocate that we `obfucate`/`sanitize` by default.

**Comment by cheempz** [comment]
Regarding 
> The appropriate place to do sanitization is in the collector, or more generally out of process. 

that disagrees with a recent [separate discussion](https://github.com/open-telemetry/opentelemetry.io/pull/3309#discussion_r1337501757), which is the case with our product--telemetry is exported directly to our platform and doesn't go though a collector.  

OTel [JS](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1237) and [Java](https://opentelemetry.io/docs/instrumentation/java/automatic/agent-config/#db-statement-sanitization) obfuscate/sanitize by default, so there's a bit of precedence despite the much needed spec clarification/configuration naming being discussed :)

**Comment by ericmustin** [comment]
Indeed, I disagree with the spec discussion. If I'm not
Mistaken, Both github and Shopify have introduced regressions into production by enabling obfuscation in app ootb. Perhaps benchmarking would be useful here, but im dubious of defaulting to obfuscation. The chief concern of observability instrumentation should be minimizing the observer effect as much as possible, and I feel we're disregarding that for what's effectively product/vendor developer experience preferences

**Comment by arielvalentin** [comment]
For context, at GH we've accepted the trade off so we obfuscate by default. It's slow but we mitigate it with the character limit setting.

**Comment by ericmustin** [comment]
just circling back, if others feel `obfuscate` is the appropriate default, then I won't stand in the way here.

**Review by arielvalentin — COMMENTED** [review]
Is this all of the data store drivers? Or do the others already have obfuscate by default?

**Comment by xuan-cao-swi** [comment]
> Is this all of the data store drivers? Or do the others already have obfuscate by default?

AFAIK, I think these three gems are the last to set obfuscate to default

**Review by arielvalentin — APPROVED** [review]

File: instrumentation/pg/test/opentelemetry/instrumentation/pg/instrumentation_test.rb [hunk 1]
```diff
@@ -51,7 +51,8 @@
     let(:user) { ENV.fetch('TEST_POSTGRES_USER', 'postgres') }
     let(:dbname) { ENV.fetch('TEST_POSTGRES_DB', 'postgres') }
     let(:password) { ENV.fetch('TEST_POSTGRES_PASSWORD', 'postgres') }
-
+    let(:config) { { db_statement: :include } }
+
```
Comment (arielvalentin) [inline instrumentation/pg/test/opentelemetry/instrumentation/pg/instrumentation_test.rb:55]:
```suggestion

```

**Review by ericmustin — APPROVED** [review]
So it goes

**Comment by arielvalentin** [comment]
@xuan-cao-swi just one more favor to ask before I merge. 

Please update the commit messages to use conventional commits:

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/CONTRIBUTING.md#use-conventional-commit-messages

This is something that I think requires a Minor release, so please using the `BREAKING CHANGE` format in the message, e.g. 

`feat!: Set db.statement option to obfuscate by default`

**Comment by xuan-cao-swi** [comment]
> @xuan-cao-swi just one more favor to ask before I merge.
> 
> Please update the commit messages to use conventional commits:
> 
> https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/CONTRIBUTING.md#use-conventional-commit-messages
> 
> This is something that I think requires a Minor release, so please using the `BREAKING CHANGE` format in the message, e.g.
> 
> `feat!: Set db.statement option to obfuscate by default`

Changed!

**Pull request merged**